src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorTest.cs

```diff
@@ -28,173 +28,283 @@ public class TimeSyncBehaviorTest
+                Assert.True(used, $"index: {i}");
+            }
+
+            diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - 800*1000) < TimeEpsilonMs, $"should be 800ms because of outbound preference. Actual: {diff.TotalMilliseconds}ms");
+        }
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// forgets old samples if it has reached predefined limits.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeInboundSampleSet_ForgetsOldSamples()
+        {
+            var dateTimeProvider = new DateTimeProvider();
+            TimeSyncBehaviorState state = this.CreateTimeSyncBehaviorState(dateTimeProvider);
+
+            // add max inbound samples
+            for (int i = 0; i < TimeSyncBehaviorState.MaxInboundSamples; i++)
+            {
+                bool used = state.AddTimeData(new IPAddress(i), TimeSpan.FromSeconds(400), true);
                 Assert.True(used);
-                if (i >= outboundSamples - TimeSyncBehaviorState.MaxOutboundSamples) outSamples.Add(-i * 1000);
             }
 
-            var allSamples = new List<int>(inSamples);
-            for (int i = 0; i < TimeSyncBehaviorState.OutboundToInboundWeightRatio; i++)
-                allSamples.AddRange(outSamples);
+            // add enough outbound samples to get an adjustment - and across an even range
+            Assert.True(state.AddTimeData( new IPAddress(1), TimeSpan.FromSeconds(-500), false));
+            Assert.True(state.AddTimeData( new IPAddress(2), TimeSpan.FromSeconds(-250), false));
+            Assert.True(state.AddTimeData( new IPAddress(3), TimeSpan.FromSeconds(250), false));
+            Assert.True(state.AddTimeData( new IPAddress(4), TimeSpan.FromSeconds(500), false));
+
+            TimeSpan diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - 250*1000) < TimeEpsilonMs, $"should be 250ms because leaning towards the inbound value of 400ms, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+
+            // add another batch of inbound with a greater offset than highest outbound
+            for (int i = TimeSyncBehaviorState.MaxInboundSamples; i < TimeSyncBehaviorState.MaxInboundSamples*2; i++)
+            {
+                Assert.True(state.AddTimeData(new IPAddress(i*2), TimeSpan.FromSeconds(-1000), true), $"index: {i}");
+            }
+
+            diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - (-250*1000)) < TimeEpsilonMs, $"should be -250ms because the new inbound replace the old and swing the median closer to the new -1000ms values, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
```

Comment:
should->Should
replace->replaces
swing->swings

Reply:
check other strings too above and make them consistent

Reply:
replace and swing were grammatically correct, but I will add the word samples after inbound to emphasise that "inbound" meant "inbound samples" collection noun.